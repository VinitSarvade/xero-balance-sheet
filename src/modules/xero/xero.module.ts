export type {
  BalanceSheetReport,
  BalanceSheetRow,
  HeaderRow,
  SectionRow,
  SectionData,
  Cell,
} from "./xeroReport.dto";
export { getReport } from "./xeroReport.service";
