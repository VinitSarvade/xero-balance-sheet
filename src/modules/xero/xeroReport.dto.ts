export interface BalanceSheetReport {
  ReportID: string;
  ReportName: string;
  ReportTitles: string[];
  ReportDate: string;
  Rows: BalanceSheetRow[];
}

export type BalanceSheetRow = SectionRow | HeaderRow;

export interface HeaderRow {
  RowType: "Header";
  Cells: Cell[];
}

export interface Cell {
  Value: string;
}

export interface SectionRow {
  RowType: "Section";
  Title: string;
  Rows: SectionData[];
}

export type SectionData = DataRow | SummaryRow;

export interface DataRow {
  RowType: "Row";
  Cells: Cell[];
}

export interface SummaryRow {
  RowType: "SummaryRow";
  Cells: Cell[];
}
