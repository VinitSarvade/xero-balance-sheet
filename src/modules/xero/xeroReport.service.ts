import { BalanceSheetReport } from "./xeroReport.dto";

const BALANCE_SHEET_API = `${process.env.XERO_API_BASE_URL}/api.xro/2.0/Reports/BalanceSheet`;

export async function getReport(): Promise<BalanceSheetReport> {
  const response = await fetch(BALANCE_SHEET_API, {
    headers: {
      // assuming authentication is done by API token
      Authorization: `Bearer ${process.env.XERO_API_TOKEN ?? ""}`,
    },
    cache: "force-cache",
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  const result = await response.json();

  return result.Reports[0];
}
