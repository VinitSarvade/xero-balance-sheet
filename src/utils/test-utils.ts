import {
  format as prettyFormat,
  plugins as prettyFormatPlugins,
} from "pretty-format";
import { http, HttpResponse } from "msw";
import { BalanceSheetReport } from "@/modules/xero/xeroReport.dto";

export function format(value: any) {
  return prettyFormat(value, {
    escapeRegex: true,
    indent: 2,
    plugins: [
      prettyFormatPlugins.ReactTestComponent,
      prettyFormatPlugins.ReactElement,
      prettyFormatPlugins.DOMElement,
      prettyFormatPlugins.DOMCollection,
      prettyFormatPlugins.Immutable,
      prettyFormatPlugins.AsymmetricMatcher,
    ],
    printFunctionName: false,
  });
}

export const BalanceSheetMock = {
  ReportID: "report-id",
  ReportName: "Balance Sheet",
  ReportDate: "2022-01-01",
  ReportTitles: ["Balance Sheet", "Test Company"],
  Rows: [
    {
      RowType: "Header",
      Cells: [{ Value: "" }, { Value: "2023" }, { Value: "2024" }],
    },
    {
      RowType: "Section",
      Title: "Assets",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Bank",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "My Bank Account" },
            { Value: "126.70" },
            { Value: "99.60" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business" },
            { Value: "92911.00" },
            { Value: "92911.00" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business 1" },
            { Value: "11039.00" },
            { Value: "11039.00" },
          ],
        },
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Bank" },
            { Value: "104076.70" },
            { Value: "104049.60" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "Liabilities",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Current Liabilities",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "Car Loan" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "",
      Rows: [
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Current Liabilities" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
  ],
} as BalanceSheetReport;

export function getHandlers() {
  return [
    http.get(`${process.env.BASE_URL}/api/xero/balance-sheet`, () => {
      return HttpResponse.json(BalanceSheetMock);
    }),
  ];
}
