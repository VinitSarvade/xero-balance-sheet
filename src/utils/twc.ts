import { createTwc } from "react-twc";
import { cn } from "./cn";

export const twc = createTwc({ compose: cn });
