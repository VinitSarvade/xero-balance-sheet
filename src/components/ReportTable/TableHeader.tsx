import { HeaderRow } from "@/modules/xero/xero.module";

interface TableHeaderProps {
  header: HeaderRow;
}

export default function TableHeader({ header }: TableHeaderProps) {
  return (
    <thead>
      <tr>
        {header.Cells.map((cell) => (
          <th key={`header-${cell.Value}`}>{cell.Value}</th>
        ))}
      </tr>
    </thead>
  );
}
