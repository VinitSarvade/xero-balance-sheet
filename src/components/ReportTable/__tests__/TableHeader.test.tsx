import { expect, describe, it } from "bun:test";
import { render } from "@testing-library/react";
import { HeaderRow } from "@/modules/xero/xero.module";
import TableHeader from "../TableHeader";
import { format } from "@/utils/test-utils";

const header: HeaderRow = {
  RowType: "Header",
  Cells: [{ Value: "Column 1" }, { Value: "Column 2" }, { Value: "Column 3" }],
};

describe("TableHeader", () => {
  it("renders table header cells correctly", () => {
    // Table header cannot be rendered without a table,
    // so we wrap it in a table element
    const { getAllByRole } = render(
      <table>
        <TableHeader header={header} />
      </table>,
    );
    const headerCells = getAllByRole("columnheader");
    expect(headerCells).toHaveLength(3);
    expect(headerCells[0].textContent).toBe("Column 1");
    expect(headerCells[1].textContent).toBe("Column 2");
    expect(headerCells[2].textContent).toBe("Column 3");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <table>
        <TableHeader header={header} />
      </table>,
    );

    expect(format(asFragment())).toMatchSnapshot();
  });

  it("renders no header cells when Cells array is empty", () => {
    const header: HeaderRow = { RowType: "Header", Cells: [] };
    const { queryAllByRole } = render(
      <table>
        <TableHeader header={header} />
      </table>,
    );
    const headerCells = queryAllByRole("columnheader");
    expect(headerCells).toHaveLength(0);
  });
});
