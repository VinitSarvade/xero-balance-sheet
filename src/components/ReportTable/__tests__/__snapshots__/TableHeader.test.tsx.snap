// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader matches snapshot 1`] = `
"<DocumentFragment>
  <table>
    <thead>
      <tr>
        <th>
          Column 1
        </th>
        <th>
          Column 2
        </th>
        <th>
          Column 3
        </th>
      </tr>
    </thead>
  </table>
</DocumentFragment>"
`;
